HPCC-16891 Utilize hostname if available when constructing conn URL

rpastrana opened the pull request:
- Utilizes SmartSocketEndpoing to fetch hostname

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana committed:
HPCC-16891 Utilize hostname if available when constructing conn URL

- Utilizes SmartSocketEndpoing to fetch hostname

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please review this change.

HPCCSmoketest commented:
Automated Smoketest
Sha: da6ced2894ae67eadb614963b3891de56affaa0a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.942 seconds |

Install hpccsystems-platform-community_6.2.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -431,7 +432,8 @@ void EsdlServiceImpl::configureUrlMethod(const char *method, IPropertyTree &entr
     entry.setProp("@path", path);
 
-    Owned<ISmartSocketFactory> sf = createSmartSocketFactory(iplist, true);
+    Owned<ISmartSocketFactory> sf = new CSmartSocketFactory(iplist, true);
```
Why do you make this change?

rpastrana replied:
To ensure we're instantiating a SmartSocket, we need to use the smartsocket smartendpoint to use access the hostname (if one exists) the ISmarSocketFactory interface is not aware of the smartendpoints. We really didn't need to use a factory pattern for these connections to begin with.

rpastrana replied:
@wangkx let me know what you think.

wangkx replied:
The create...() does give you an instance of SmartSocket. We use many create...() functions to create a class instance. The create...() functions may 'encapsulate' some pre-processes with the instance creation. It looks wired to change from an create...() to new...() although the code does work...

rpastrana replied:
There wasn't an actual need to follow the factory pattern here (although factories can be very elegant in some cases). We need an instance of SmartSocket, the create method currently only instantiates a SmartSocket, in the future it could instantiate others.

richardkchapman replied:
That seems like an argument for leaving the code the way it was to me. If createSmartSocketFactory creates something other than a CSmartSocketFactory, it's because something (presumably a global setting) wanted Smart Sockets to use a different implementation (maybe creating an auditing version, for example). You would be bypassing such a mechanism here by making that change.

wangkx reviewed (commented):
@rpastrana just a question.

wangkx reviewed (approved):
Ok, the name 'smartSocketfactory' is mis-leading. But, we do not want to change it here.
@rpastrana the code looks fine.

richardkchapman commented:
@rpastrana Typo in commit message

richardkchapman commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -866,19 +869,25 @@ void EsdlServiceImpl::generateTargetURL(IEspContext & context,
 
-    ISmartSocketFactory *conn = connMap.getValue(name);
-    if (!conn)
+    CSmartSocketFactory *sconn = static_cast<CSmartSocketFactory*>(connMap.getValue(name));
```
I don't like this - it's a bad assumption (see comment above)

rpastrana replied:
connMap only contains instances of CSmartSocketFactory. however the interface change will make this static cast go away.

richardkchapman commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -866,19 +869,25 @@ void EsdlServiceImpl::generateTargetURL(IEspContext & context,
-    ep = conn->nextEndpoint();
-
-    ep.getUrlStr(url);
+    SmartSocketEndpoint * sep = sconn->nextSmartEndpoint();
```
I'm guessing you made these changes because nextSmartEndpoint is not in the ISmartSocketFactory interface? The correct fix would have been to add it to the interface (or to a derived interface)...

rpastrana replied:
Correct, this is the reason I instantiated the SmartSocket, I thought about adding it to the ISmartSocketFactory but it seemed like an implementation specific feature. The reason I wanted the SEP was to access the hostname. I'll add a method to query the next available hostname if available.

rpastrana committed:
HPCC-16891 Add implementation independent method to ISmartSocket interface

- ISmartSocket requires implementation of getUrlStr
- Esdl binding utilizes IsmartSocket::getUrlStr

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@richardkchapman made changes based on your review fccfbcb

HPCCSmoketest commented:
Automated Smoketest
Sha: fccfbcbc5d3020368d52229bc165e526c7312375
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.234 seconds |

Install hpccsystems-platform-community_6.2.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
Looks sensible. @wangkx can you recheck as you understand this code better than I do

wangkx commented:
@rpastrana 2 questions.

rpastrana commented on system/jlib/jsmartsock.cpp:
```diff
@@ -428,6 +428,34 @@ void CSmartSocketFactory::setStatus(SocketEndpoint &ep, bool status)
         ss->status=status;
 }
 
+StringBuffer & CSmartSocketFactory::getUrlStr(StringBuffer &url, bool useHostName, unsigned dnsIntervalSecs)
```
good catch

rpastrana commented on system/jlib/jsmartsock.cpp:
```diff
@@ -428,6 +428,34 @@ void CSmartSocketFactory::setStatus(SocketEndpoint &ep, bool status)
+
+		if(useHostName)
+		{
+			ep = sep->ep;
```
Because we're constructing the URL based on the hostname, but perhaps the else portion does require a checkhost()

rpastrana replied:
confirmed the checkHost() will attempt a hostname to ip resolution as long as it hasn't been done within the specified dnsinterval.

richardkchapman commented:
@rpastrana Looks like this is back with you

rpastrana commented:
@wangkx please review latest changes.

HPCCSmoketest commented:
Automated Smoketest
Sha: 1943f4c2d95c4fefd94deb513da8188c39a9fd5a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.666 seconds |

Install hpccsystems-platform-community_6.2.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented on system/jlib/jsmartsock.cpp:
```diff
@@ -428,6 +428,34 @@ void CSmartSocketFactory::setStatus(SocketEndpoint &ep, bool status)
+					url.append(':').append((unsigned)ep.port);
+			}
+			else
+				ep.getUrlStr(url);
```
Do you need to call checkHost() before this line?

rpastrana committed:
HPCC-16891 Code review changes

- Removed unused input param
- Simplified logic

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx simplified my logic in order to address the concern you mentioned. Please confirm

HPCCSmoketest commented:
Automated Smoketest
Sha: 75cd6c29ff0c12aa3b6351d27549c8a3062d4fbb
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.444 seconds |

Install hpccsystems-platform-community_6.2.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented:
@rpastrana looks fine now.

rpastrana commented:
@richardkchapman please merge
